File: original_code/AR_Gamepad_to_UDP.py
# Joystick and Socket handling written by ChatGPT
# Conceptualized, corrected, and modified by Adam Kemp, 2024
# V1 - Completed basic joystick to UDP funcitonality

from numpy import interp # So we can use the Arduino's "map()" command. Not sure how else to do it in Python.
import time # Allows us to slow down the transmit rate

# Package needed to handle joysticks
import pygame

# Package needed to handle UDP communication
import socket

# Initialize pygame
pygame.init()
pygame.joystick.init()

# The sample rate for joystick data
samplerate = 0.1 # In s

# Check to see if any joysicks are available
if(pygame.joystick.get_count() == 0):
    # If not, close the program
    print("\nNo controllers have been detected! What's up with that? Exiting...")
    pygame.joystick.quit()
    pygame.quit()
    quit()

# Create a list of joysticks
print("\nAvailable Controllers: ")
for i in range (pygame.joystick.get_count()):
    tempJoy = pygame.joystick.Joystick(i)
    tempJoy.init()
    print(str(i) + ": " + str(tempJoy.get_name()))
    tempJoy.quit()

# Initialize joystick
sel = input("\nEnter the number of the controller you would like to use: ")
joystick = pygame.joystick.Joystick(int(sel))
joystick.init()

# Initialize UDP socket
UDP_IP = input("Enter the destination IP address: ")  # Change this to the IP address of the receiver
UDP_PORT = int(input("Enter the destination port number: "))
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

print("Starting... Press ctrl+c to quit.")
time.sleep(1)

try:
    while True:
        # Get events
        pygame.event.pump()

        # Get joystick axes and buttons, then map each range to something the Arduino can use, like 0-255
        axes = [int(interp(joystick.get_axis(i),[-1.0,1.0],[0,255])) for i in range(joystick.get_numaxes())]
        # Get button values. These are binary, so no mapping is needed
        buttons = [joystick.get_button(i) for i in range(joystick.get_numbuttons())]
        
        # Combine axes and buttons into a single list
        data = axes + buttons
        data.append(0xff) # Add a null to the end of the list
        print(data)

        # Convert data to bytes
        data_bytes = bytes(data)

        # Send data over UDP
        sock.sendto(data_bytes, (UDP_IP, UDP_PORT))

        #Set the sample rate
        time.sleep(samplerate)

except KeyboardInterrupt:
    print("Exiting...")
    joystick.quit()
    pygame.quit()

File: AR_Gamepad_to_UDP.js
// Joystick and Socket handling - Node.js version
// Inspired by the Python script and adapted for Node.js

const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
});

const gamepad = require('gamepad');
const dgram = require('dgram');

// Initialize gamepad library
gamepad.init();

// Function to map a value from one range to another
const mapRange = (value, inMin, inMax, outMin, outMax) => {
  return (value - inMin) * (outMax - outMin) / (inMax - inMin) + outMin;
};

// Check for connected gamepads
const getGamepadCount = () => {
  return gamepad.numDevices();
};

// Get gamepad details
const getGamepadInfo = (index) => {
  return gamepad.deviceAtIndex(index);
};

// Main function
const main = async () => {
  // Check if any gamepads are connected
  if (getGamepadCount() === 0) {
    console.log("\nNo controllers have been detected! What's up with that? Exiting...");
    process.exit();
  }

  // List available gamepads
  console.log("\nAvailable Controllers: ");
  for (let i = 0; i < getGamepadCount(); i++) {
    const gamepadInfo = getGamepadInfo(i);
    console.log(`${i}: ${gamepadInfo.description}`);
  }

  // Get user input for desired gamepad
  const selectedGamepadIndex = await new Promise((resolve) => {
    readline.question('\nEnter the number of the controller you would like to use: ', resolve);
  });

  // Initialize UDP socket
  const udpSocket = dgram.createSocket('udp4');

  // Get user input for destination IP and port
  const udpIp = await new Promise((resolve) => {
    readline.question('Enter the destination IP address: ', resolve);
  });
  const udpPort = await new Promise((resolve) => {
    readline.question('Enter the destination port number: ', resolve);
  });

  console.log("Starting... Press Ctrl+C to quit.");

  // Sample rate for gamepad data
  const sampleRate = 100; // milliseconds

  // Gamepad data processing loop
  setInterval(() => {
    // Read gamepad state
    gamepad.processEvents();
    const gamepadState = getGamepadInfo(selectedGamepadIndex).state;

    // Extract axes and map their values to 0-255 range
    const axes = gamepadState.axes.map((axisValue) => Math.round(mapRange(axisValue, -1, 1, 0, 255)));

    // Extract button values (already binary)
    const buttons = gamepadState.buttons.map((button) => button.pressed ? 1 : 0);

    // Combine axes and buttons into a single data array
    const data = axes.concat(buttons);
    data.push(0xff); // Add null terminator

    console.log(data);

    // Send data over UDP
    const message = Buffer.from(data);
    udpSocket.send(message, 0, message.length, udpPort, udpIp);
  }, sampleRate);

  // Handle Ctrl+C to exit
  process.on('SIGINT', () => {
    console.log("Exiting...");
    process.exit();
  });
};

// Run the main function
main();